### PR TITLE
Plugins: CLI Plugin Manager SubCommand

### PR DESCRIPTION
**Pull request by clintoncwolfe**
This PR expands the UX of inspec, by adding a core plugin named `inspec-plugin-manager-cli`.  The plugin is responsible for managing the installation of InSpec and Train user-install plugins.

One subcommand, and five subsubcommands, are added:

* `inspec plugin list` - list installed plugins
* `inspec plugin search` - query rubygems.org for available plugins
* `inspec plugin install` - install a local or remote gem, or a reference to a local source tree, as a plugin
* `inspec plugin update` - upgrade installed plugins
* `inspec plugin uninstall` - remove an installed plugin

On the filesystem, `~/.inspec/plugins.json` is used as a statefile, and the gem activity is privately managed under `~/.inspec/gems`.  `inspec-plugin-manager-cli` promises to enforce a coherent, clean set of gems in that area, so long as the user does not meddle directly in it. Violations of that promise are to be considered bugs.

CLI usage help is included, but perhaps uneven; some commands have a `long_desc`, while others do not, perhaps due to obviousness.

Functional tests are extensive, but still likely incomplete; UX messaging and exit status are key focuses, under commonly expected conditions. It is likely that users will find new and interesting edge cases.

The Chef InSpec engineering team intends this PR to mark the 2.2 -> 2.3 version bump.

attn: @mattray

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ AllCops:
   - Gemfile
   - Rakefile
   - 'test/**/*'
+  - 'lib/plugins/*/test/**/*'
   - 'examples/**/*'
   - 'vendor/**/*'
   - 'lib/plugins/inspec-*/test/**/*'
```
Comment (jquick) [inline .rubocop.yml]: Can you remove this line? Yours is superior.

File: docs/plugins.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * DSL extensions at the file, control, or test level
+ * attribute fetchers to allow reading InSpec attributes from new sources (for example, a remote, encrypted key-value store)
+ *
+
+## How do I find out which plugins are available?
+
+The InSpec CLI can tell you which plugins are available:
+
+```
+you@machine $ inspec plugin search
```
Comment (jquick) [inline docs/plugins.md]: Im not sure we do this `you@machine` anywhere else. I think just leaving that off or just `$` is fine. You may also want to ```Bash header it for format. Not a blocker.

**Comment by chris-rock** [comment]
If I point inspec plugin install to a root directory of a gem, I am getting the following error:
image
Based on the output it is also unclear how to fix the issue. Seems like this is affecting all inspec execution, since the failure effects all commands. Should include a verification step during installation and revert the installation if we could not load it?

I removed ~/.inspec/plugins.json manually to fix it locally

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+      #                       inspec plugin install
+      #==================================================================#
+      desc 'install [-v VERSION] PLUGIN', 'Installs a plugin from rubygems.org, a gemfile, or a path to local source.'
+      long_desc 'PLUGIN may be the name of a gem on rubygems.org that begins with inspec- or train-.  PLUGIN may also be the path to a local gemfile, which will then be installed like any other gem.  Finally, if PLUGIN is a path ending in .rb, it is taken to be a local file that will act as athe entry point for a plugin (this mode is provided for local plugin development).  Exit codes are 0 on success, 2 if the plugin is already installed, and 1 if any other error occurs.'
+      option :version, desc: 'When installing from rubygems.org, specifies a specific version to install.', aliases: [:v]
+      def install(plugin_id_arg)
+        case plugin_id_arg
+        when /\.gem$/ # Does it end in .gem?
+          install_from_gemfile(plugin_id_arg)
+        when %r{[\/\\]} # Does the argument have a slash?
```
Comment (jquick) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb]: Does this mean if you want to install from the same folder you would have to do `install ./my_plugin`? Maybe we could do a quick `|| File.exist?` at this level also?

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+      #--------------------------
+      #        update
+      #--------------------------
+      desc 'update PLUGIN', 'Updates a plugin to the latest from from rubygems.org'
+      long_desc 'PLUGIN may be the name of a gem on rubygems.org that begins with inspec- or train-.  Exit codes are 0 on success, 2 if the plugin is already up to date, and 1 if any other error occurs.'
+      def update(plugin_name)
+        pre_update_versions = installer.list_installed_plugin_gems.select { |spec| spec.name == plugin_name }.map { |spec| spec.version.to_s }
+        old_version = pre_update_versions.join(', ')
+
+        update_preflight_check(plugin_name, pre_update_versions)
```
Comment (jquick) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb:100]: love the name of this method 🛫

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+      def install_from_path(path)
+        unless File.exist? path
+          puts(red { 'No such source code path ' } + path + ' - installation failed.')
+          exit 1
+        end
+
+        plugin_name = File.basename(path)
+        check_plugin_name(plugin_name, 'installation')
+
+        installer.install(plugin_name, path: path)
```
Comment (jquick) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb]: Should we expand_path here? What if they install from the same folder but then later run somewhere else? You may of already covered this and I missed it.

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/plugin.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+module InspecPlugins
```
Comment (jquick) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/plugin.rb:1]: Do you need this file? This seems to be the same as `lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli.rb`

**Review by jquick — COMMENTED** [review]
Thanks @clintoncwolfe. This looks really good. Fantastic test coverage.

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/plugin.rb (hunk 6 above)
Comment (clintoncwolfe) [reply lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/plugin.rb:1]: Normally I would compress that to InspecPlugins::PluginManager , but under unit testing, both plugins.rb and inspec-plugin-manager-cli.rb are loaded in isolation.  So, they both must declare that module, to a void a NameError.

It's harmless to do twice, as the second one to execute will simply re-open the module.

**Comment by clintoncwolfe** [comment]
The code around path install is indeed very brittle.  We can expand the hueristics more ("it should have a slash, and then I'll assume that worked") and run verification that the identified entry point exists and actually can be loaded, to result in a v2 plugin.

I sweetened the UX and error handling on the path-mode-install less than other modes, because the envisioned user mode would typically be remote gem install, with only developers using path-mode for local development.

Agree though, it's too brittle.  I'll add more test cases and safeties.

**Comment by clintoncwolfe** [comment]
@chris-rock path install can now handle a variety of paths, including the root of a plugin project.  See updated functional test.

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+      rescue Inspec::Plugin::V2::SearchError => ex
+        Inspec::Log.error ex.message
+        exit 1
+      end
+
+      #==================================================================#
+      #                       inspec plugin install
+      #==================================================================#
+      desc 'install [-v VERSION] PLUGIN', 'Installs a plugin from rubygems.org, a gemfile, or a path to local source.'
+      long_desc 'PLUGIN may be the name of a gem on rubygems.org that begins with inspec- or train-.  PLUGIN may also be the path to a local gemfile, which will then be installed like any other gem.  Finally, if PLUGIN is a path ending in .rb, it is taken to be a local file that will act as athe entry point for a plugin (this mode is provided for local plugin development).  Exit codes are 0 on success, 2 if the plugin is already installed, and 1 if any other error occurs.'
```
Comment (miah) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb]: Not a blocker, but can we wrap this and other lines at ~80 chars?

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+        puts(red { 'Unrecognizable plugin structure' } + " - #{parts[2]} - When installing from a path, please provide the path of the entry point file - installation failed.")
+        exit 1
+      end
+
+      def install_from_path__probe_load(entry_point, plugin_name)
+        # Brazenly attempt to load a file, and see if it registers a plugin.
+        begin
+          require entry_point
+        rescue LoadError => ex
+          puts(red { 'Plugin contains errors' } + " - #{plugin_name} - Encountered errors while trying to test laod the plugin entry point, resolved to #{entry_point} - installation failed")
```
Comment (miah) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb]: s/laod/load

**Review by miah — APPROVED** [review]
Only a few minor quibbles. No blockers. This looks great! The method names lead to very readable conditionals. =) Thanks @clintoncwolfe

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 9]
```diff
@@ -165,15 +165,68 @@ def install_from_path(path)
...
+        parts = [
+          given.parent.parent.basename,
+          given.parent.basename,
+          given.basename('.rb'),
+          given.extname,
+        ].map(&:to_s)
+
+        # Simplest case: it was a full entry point, as presented.
+        # /home/you/projects/inspec-something/lib/inspec-something.rb
+        #   parts index:           ^0^        ^1^      ^2^        ^3^
```
Comment (jquick) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb:225]: Really nice documentation here!

**Review by jquick — APPROVED** [review]
Thanks @clintoncwolfe, looking really good. Love the path additions.

**Comment by jquick** [comment]
DCO bot is confused with the amount of commits on this. This is signed off accordingly. View as all checks passing.
